jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/JibExtension.java

```diff
@@ -221,6 +227,17 @@ public String getContainerizingMode() {
     return property != null ? property : containerizingMode.get();
   }
 
+  @Input
+  @Optional
+  public Property<String> getConfigurationName() {
+    return configurationName;
+  }
+
+  public String readConfigurationName() {
+    String property = System.getProperty(PropertyNames.CONFIGURATION_NAME);
+    return property != null ? property : configurationName.get();
+  }
+
```

Comment:
Because the getter (and the get and the read function have a different semantic meaning) has a special meaning in kotlin, which is used for the gradle kotlin DSL (which is a official, fully supported option besides the old groovy DSL). For this property I changed it from the current model of bare getters and setters to expose the `Property` itself (This would be good for all options, but because it's not backwards compatible it seems like a good idea to only introduce it for new properties). This is a best practice, as it allows for lazily setting the value of the option for example (this also applies to the groovy DSL).

The getter for the property has to be named like a standard getter for kotlin to allow the conversation to a synthetic property like this:
```kotlin
jib {
    configurationName.set("value")
}
```
If the getter is name differently than the standard getter this happens:
```kotlin
jib {
    alternativeGetter().set("value")
}
```
which is probably not desirable.

One alternative that might be possible is the usage of a `public final` property, I will investigate that and report back on both solutions' compatibility with both the groovy and the kotlin DSL, but in either case the getter seems not a good idea in that case.

One additional option that's independent from the two above is to replace all instances of the read function with a `getConfigurationName().get()` but that is missing the default value default (which might not be needed though, I'll investigate, whether it can ever return `null`).

Reply:
One note about the exposing of `Property`: The recent API additions of feels also seem to heavily favor this style ([look here](https://docs.gradle.org/current/userguide/toolchains.html#sec:consuming) and switch to the kotlin DSL and notice the `languageVersion.set()`)

Reply:
I appreciate your detailed explanation. I can now see why you are trying to expose the Property using the getter (since that seems to be the convention for lazy configuration). The part that is a little confusing is the call to `readConfigurationName()` whenever `configurationName` is referenced. I'm leaning more towards your third option, where we use `getConfigurationName().get()` instead of `readConfigurationName()`  which also aligns with the examples in https://docs.gradle.org/current/userguide/implementing_gradle_plugins.html. But yeah, we need to double-check if it is possible for it to return `null`.

Reply:
It'd be really nice if we could expose `Property<String>` in general, but there's a catch: I believe the reason that we factor in reading system properties in the getters marked with `@Input` is to correctly make Gradle check UP-TO-DATE. That is, Jib doesn't only read input values from `build.gradle` but also from other external source. For example, let's say you ran Jib once and a subsequent Gradle build will skip running the Jib task if there's no change to `build.gradle`. However, if you change some option on the command-line (e.g., with `-Djib.configurationName=...`), we want Gradle to re-run Jib even if nothing has changed in `build.gradle`.

Reply:
[To correct my comment above] After discussing this a bit and given @chanseokoh's comment, the last point in my comment about checking for the `null` value is not really valid.
But as mentioned above, due to the uniqueness of our situation and how system properties take precedence over values from `build.gradle` in jib, it still seems to be the case that we need to have this logic about system properties in the getter.

Reply:
Yeah right I misread the code, it's not a null-check it's a system.property override

Reply:
So the `@Input` should be moved down

Reply:
> So the `@Input` should be moved down

I don't think you can.
```
   > Warning: Type 'JibExtension': non-property method 'readConfigurationName()' should not be annotated with: @Input, @Optional.
```

Reply:
The problem I see otherwise is that changes of the property won't be reflected by the up-to-date status. Maybe we should stop the system

Reply:
It should have been nice if we were able to add `@Input` to `readConfigurationName()`, but looks like `java-gradle-plugin` doesn't allow it probably because of the weird naming. (OTOH, certainly we can add `@Input` to `getConfigurationName()`.) And in order for Gradle to properly work with the UP-TO-DATE check in the present of our support for system properties, I think the best course of action is to keep `@Input` on `getConfigurationName()` while the method should factor in the system property. I did confirm that the UP-TO-DATE status is correctly reflected based on the return value of a getter marked with `@Input`. We can still hold and accept a value as `Property`.

Reply:
That does break compatibility with the kotlin DSL though and that's the dilemma, so it seems like it's either proper, property based configuration, with maybe defaults based on a system property or a system property override.